crypto/hpke.c

```diff
@@ -0,0 +1,2730 @@
+            concat_offset+=strlen(HPKE_SEC41LABEL);
+            if (concat_offset>=HPKE_MAXSIZE) { erv=__LINE__; goto err; }
+            labeled_ikm[concat_offset]=(suite.kem_id/256)%256;
+            concat_offset+=1;
+            if (concat_offset>=HPKE_MAXSIZE) { erv=__LINE__; goto err; }
+            labeled_ikm[concat_offset]=suite.kem_id%256;
+            concat_offset+=1;
+            if (concat_offset>=HPKE_MAXSIZE) { erv=__LINE__; goto err; }
+            memcpy(labeled_ikm+concat_offset,label,labellen);
+            concat_offset+=labellen;
+            if (concat_offset>=HPKE_MAXSIZE) { erv=__LINE__; goto err; }
+            memcpy(labeled_ikm+concat_offset,ikm,ikmlen);
+            concat_offset+=ikmlen;
+            if (concat_offset>=HPKE_MAXSIZE) { erv=__LINE__; goto err; }
+            labeled_ikmlen=concat_offset;
+            break;
+
+        case HPKE_5869_MODE_FULL:
+            concat_offset=0;
+            memcpy(labeled_ikm,HPKE_VERLABEL,strlen(HPKE_VERLABEL));
+            concat_offset+=strlen(HPKE_VERLABEL);
+            if (concat_offset>=HPKE_MAXSIZE) { erv=__LINE__; goto err; }
+            memcpy(labeled_ikm+concat_offset,
+                    HPKE_SEC51LABEL,strlen(HPKE_SEC51LABEL));
+            concat_offset+=strlen(HPKE_SEC51LABEL);
+            if (concat_offset>=HPKE_MAXSIZE) { erv=__LINE__; goto err; }
+            labeled_ikm[concat_offset]=(suite.kem_id/256)%256;
+            concat_offset+=1;
+            if (concat_offset>=HPKE_MAXSIZE) { erv=__LINE__; goto err; }
+            labeled_ikm[concat_offset]=suite.kem_id%256;
+            concat_offset+=1;
+            if (concat_offset>=HPKE_MAXSIZE) { erv=__LINE__; goto err; }
+            labeled_ikm[concat_offset]=(suite.kdf_id/256)%256;
+            concat_offset+=1;
+            if (concat_offset>=HPKE_MAXSIZE) { erv=__LINE__; goto err; }
+            labeled_ikm[concat_offset]=suite.kdf_id%256;
+            concat_offset+=1;
+            if (concat_offset>=HPKE_MAXSIZE) { erv=__LINE__; goto err; }
+            labeled_ikm[concat_offset]=(suite.aead_id/256)%256;
+            concat_offset+=1;
+            if (concat_offset>=HPKE_MAXSIZE) { erv=__LINE__; goto err; }
+            labeled_ikm[concat_offset]=suite.aead_id%256;
+            concat_offset+=1;
+            if (concat_offset>=HPKE_MAXSIZE) { erv=__LINE__; goto err; }
+            memcpy(labeled_ikm+concat_offset,label,labellen);
+            concat_offset+=labellen;
+            if (concat_offset>=HPKE_MAXSIZE) { erv=__LINE__; goto err; }
+            if (ikmlen>0) /* added 'cause asan test */
+            memcpy(labeled_ikm+concat_offset,ikm,ikmlen);
+            concat_offset+=ikmlen;
+            if (concat_offset>=HPKE_MAXSIZE) { erv=__LINE__; goto err; }
+            labeled_ikmlen=concat_offset;
+            break;
+        default:
+            erv=__LINE__; goto err;
+    }
+
+    pctx = EVP_PKEY_CTX_new_from_name(hpke_libctx, "HKDF", NULL);
```

Comment:
Using the new EVP_KDF APIs would be preferable here, rather than using the older EVP_PKEY_derive route.

Reply:
Thanks - will take a look and get back here.

Reply:
Looks like that seems ok, at least for expands and extracts - I pushed a version of https://github.com/sftcd/happykey that does that now and it seems to work ok - no real saving in lines of code but I guess it has fewer error exits and is more in tune with what you prefer for upstream so is worthwhile. Will do some more testing etc (in particular on leaks after error exits when used with various algs and for ECH ) and then likely push that here in the not-too-distant. Thanks again for the heads-up, other similar ones welcome.

Reply:
fixed in 175ff1e (Hopefully)